// es-drager/lib/index.es.js
import { computed as K, ref as N, onMounted as me, watch as U, onBeforeUnmount as Be, nextTick as Ce, defineComponent as we, openBlock as $, createElementBlock as W, renderSlot as O, createElementVNode as j, createBlock as he, resolveDynamicComponent as Re, normalizeClass as _e, unref as S, normalizeStyle as de, withModifiers as Ne, withCtx as fe, Fragment as Se, renderList as Pe, normalizeProps as Ie, guardReactiveProps as Ke, createCommentVNode as ge } from "vue";
const $e = {
  tag: {
    type: [String, Object],
    default: "div"
  },
  resizable: {
    type: Boolean,
    default: !0
  },
  rotatable: {
    type: Boolean,
    default: !1
  },
  boundary: {
    // 边界
    type: Boolean
  },
  disabled: Boolean,
  width: {
    type: Number,
    default: 0
  },
  height: {
    type: Number,
    default: 0
  },
  left: {
    type: Number,
    default: 0
  },
  top: {
    type: Number,
    default: 0
  },
  zIndex: {
    type: Number,
    default: 0
  },
  angle: {
    type: Number,
    default: 0
  },
  color: {
    type: String,
    default: "#3a7afe"
  },
  minWidth: {
    type: Number,
    default: -1 / 0
  },
  minHeight: {
    type: Number,
    default: -1 / 0
  },
  maxWidth: {
    type: Number,
    default: 0
  },
  maxHeight: {
    type: Number,
    default: 0
  },
  aspectRatio: {
    // 缩放比例
    type: Number
  },
  selected: Boolean,
  snapToGrid: Boolean,
  gridX: {
    type: Number,
    default: 50
  },
  gridY: {
    type: Number,
    default: 50
  },
  scaleRatio: {
    type: Number,
    default: 1
  },
  disabledKeyEvent: Boolean,
  border: {
    type: Boolean,
    default: !0
    // 兼容默认
  },
  resizeList: {
    type: Array
    // 要显示的缩放handle列表，默认显示全部
  },
  equalProportion: {
    // 是否等比例缩放
    type: Boolean
  },
  checkCollision: {
    // 是否开启碰撞检查
    type: Boolean
  },
  snap: Boolean,
  // 是否开启吸附
  snapThreshold: {
    // 吸附阈值，默认10px
    type: Number,
    default: 10
  },
  markline: [Boolean, Function]
  // 辅助线
};
function te(o, n) {
  const e = (t) => {
    n && n(t), document.removeEventListener("mousemove", o), document.removeEventListener("mouseup", e), document.removeEventListener("mouseleave", e), document.removeEventListener("touchmove", o), document.removeEventListener("touchend", e);
  };
  document.addEventListener("mousemove", o), document.addEventListener("mouseup", e), document.addEventListener("mouseleave", e), document.addEventListener("touchmove", o), document.addEventListener("touchend", e);
}
function V(o) {
  let n = 0, e = 0;
  if (Ve(o)) {
    const t = o.targetTouches[0];
    n = t.pageX, e = t.pageY;
  } else
    n = o.clientX, e = o.clientY;
  return { clientX: n, clientY: e };
}
function Ve(o) {
  const n = Object.prototype.toString.call(o);
  return n.substring(8, n.length - 1) === "TouchEvent";
}
const F = (o = 0) => parseInt(o + "") + "px", H = {
  n: "top",
  s: "bottom",
  e: "right",
  w: "left",
  ne: "top-right",
  nw: "top-left",
  se: "bottom-right",
  sw: "bottom-left"
}, ee = ["n", "ne", "e", "se", "s", "sw", "w", "nw"], Ge = { n: 0, ne: 1, e: 2, se: 3, s: 4, sw: 5, w: 6, nw: 7 }, qe = {
  0: 0,
  1: 1,
  2: 2,
  3: 2,
  4: 3,
  5: 4,
  6: 4,
  7: 5,
  8: 6,
  9: 6,
  10: 7,
  11: 8
}, Fe = (o, n) => {
  const e = qe[Math.floor(o / 30)], i = (Ge[n] + e) % 8;
  return ee[i];
}, ve = (o = 0, n) => {
  let e = [];
  for (let t = 0; t < ee.length; t++) {
    const i = ee[t], [h, d] = H[i].split("-"), s = Fe(o, i), c = {
      [h]: "0%",
      cursor: s + "-resize",
      side: H[i]
    };
    if (d)
      c[d] = "0%";
    else {
      const g = ["top", "bottom"].includes(h) ? "left" : "top";
      c[g] = "50%";
    }
    n ? n.includes(H[i]) && e.push(c) : e.push(c);
  }
  return e;
}, ne = (o) => o * Math.PI / 180, Ue = (o, n) => Math.sqrt(o * o + n * n), k = (o) => Math.cos(ne(o)), L = (o) => Math.sin(ne(o)), Oe = (o, n, e, t, i, h, d) => {
  let { width: s, height: c, centerX: g, centerY: w, rotateAngle: l } = n;
  const M = s < 0 ? -1 : 1, v = c < 0 ? -1 : 1;
  switch (s = Math.abs(s), c = Math.abs(c), o) {
    case "right": {
      const u = I(s, e, h);
      s = u.width, e = u.deltaW, i ? (t = e / i, c = s / i, g += e / 2 * k(l) - t / 2 * L(l), w += e / 2 * L(l) + t / 2 * k(l)) : (g += e / 2 * k(l), w += e / 2 * L(l));
      break;
    }
    case "top-right": {
      t = -t;
      const u = I(s, e, h);
      s = u.width, e = u.deltaW;
      const r = P(c, t, d);
      c = r.height, t = r.deltaH, i && (e = t * i, s = c * i), g += e / 2 * k(l) + t / 2 * L(l), w += e / 2 * L(l) - t / 2 * k(l);
      break;
    }
    case "bottom-right": {
      const u = I(s, e, h);
      s = u.width, e = u.deltaW;
      const r = P(c, t, d);
      c = r.height, t = r.deltaH, i && (e = t * i, s = c * i), g += e / 2 * k(l) - t / 2 * L(l), w += e / 2 * L(l) + t / 2 * k(l);
      break;
    }
    case "bottom": {
      const u = P(c, t, d);
      c = u.height, t = u.deltaH, i ? (e = t * i, s = c * i, g += e / 2 * k(l) - t / 2 * L(l), w += e / 2 * L(l) + t / 2 * k(l)) : (g -= t / 2 * L(l), w += t / 2 * k(l));
      break;
    }
    case "bottom-left": {
      e = -e;
      const u = I(s, e, h);
      s = u.width, e = u.deltaW;
      const r = P(c, t, d);
      c = r.height, t = r.deltaH, i && (c = s / i, t = e / i), g -= e / 2 * k(l) + t / 2 * L(l), w -= e / 2 * L(l) - t / 2 * k(l);
      break;
    }
    case "left": {
      e = -e;
      const u = I(s, e, h);
      s = u.width, e = u.deltaW, i ? (c = s / i, t = e / i, g -= e / 2 * k(l) + t / 2 * L(l), w -= e / 2 * L(l) - t / 2 * k(l)) : (g -= e / 2 * k(l), w -= e / 2 * L(l));
      break;
    }
    case "top-left": {
      e = -e, t = -t;
      const u = I(s, e, h);
      s = u.width, e = u.deltaW;
      const r = P(c, t, d);
      c = r.height, t = r.deltaH, i && (s = c * i, e = t * i), g -= e / 2 * k(l) - t / 2 * L(l), w -= e / 2 * L(l) + t / 2 * k(l);
      break;
    }
    case "top": {
      t = -t;
      const u = P(c, t, d);
      c = u.height, t = u.deltaH, i ? (s = c * i, e = t * i, g += e / 2 * k(l) + t / 2 * L(l), w += e / 2 * L(l) - t / 2 * k(l)) : (g += t / 2 * L(l), w -= t / 2 * k(l));
      break;
    }
  }
  return {
    position: {
      centerX: g,
      centerY: w
    },
    size: {
      width: s * M,
      height: c * v
    }
  };
}, P = (o, n, e) => {
  const t = o + n;
  return t > e ? o = t : (n = e - o, o = e), { height: o, deltaH: n };
}, I = (o, n, e) => {
  const t = o + n;
  return t > e ? o = t : (n = e - o, o = e), { width: o, deltaW: n };
}, je = ({
  centerX: o,
  centerY: n,
  width: e,
  height: t,
  angle: i
}) => ({
  top: n - t / 2,
  left: o - e / 2,
  width: e,
  height: t,
  angle: i
}), Je = (o, n, e) => {
  const { width: t, height: i } = o;
  return {
    width: Math.abs(t),
    height: Math.abs(i),
    left: n - Math.abs(t) / 2,
    top: e - Math.abs(i) / 2
  };
};
function J(o, n) {
  const e = Math.abs(o) % n, t = o > 0 ? n : -n;
  let i = 0;
  return e > n / 2 ? i = t * Math.ceil(Math.abs(o) / n) : i = t * Math.floor(Math.abs(o) / n), i;
}
function Qe(o, n, e) {
  if (!o || !n)
    return !1;
  const t = R(o, e), i = R(n, e);
  return t.left < i.left + i.width && t.left + t.width > i.left && t.top < i.top + i.height && t.top + t.height > i.top;
}
const R = (o, n) => {
  var e = o.getBoundingClientRect();
  return {
    ...e,
    left: e.left / n,
    top: e.top / n,
    right: e.right / n,
    bottom: e.bottom / n,
    width: e.width / n,
    height: e.height / n
  };
}, ye = (o) => typeof o == "function";
function Ze(o, n) {
  let e = null, t = null;
  const i = K(() => o.value.offsetParent || document.body), h = K(() => R(i.value, n.scaleRatio)), d = N({ x: [], y: [] }), s = () => {
    n.markline && !ye(n.markline) && (e || (e = document.querySelector(".es-drager-markline-x") || pe("x", i.value, n.color)), t || (t = document.querySelector(".es-drager-markline-y") || pe("y", i.value, n.color)));
  }, c = (v = {}) => {
    if (n.markline) {
      if (ye(n.markline))
        return n.markline(v);
      v.left === null ? e.style.display = "none" : (e.style.left = v.left + "px", e.style.backgroundColor = n.color, e.style.display = "block"), v.top === null ? t.style.display = "none" : (t.style.top = v.top + "px", t.style.backgroundColor = n.color, t.style.display = "block");
    }
  }, g = () => {
    const v = R(o.value, n.scaleRatio), u = document.querySelectorAll(".es-drager"), r = [];
    for (let f = 0; f < u.length; f++) {
      const p = u[f];
      p !== o.value && r.push(R(p, n.scaleRatio));
    }
    d.value = He(r, v);
  }, w = () => {
    const v = {
      top: null,
      left: null,
      diffX: 0,
      diffY: 0
    }, u = R(o.value, n.scaleRatio);
    for (let r = 0; r < d.value.y.length; r++) {
      const { top: f, showTop: p } = d.value.y[r];
      if (Math.abs(f - u.top) < n.snapThreshold) {
        v.diffY = f - u.top, v.top = p - h.value.top;
        break;
      }
    }
    for (let r = 0; r < d.value.x.length; r++) {
      const { left: f, showLeft: p } = d.value.x[r];
      if (Math.abs(f - u.left) < n.snapThreshold) {
        v.diffX = f - u.left, v.left = p - h.value.left;
        break;
      }
    }
    return c(v), v;
  }, l = () => {
    c({ left: null, top: null });
  }, M = (v) => {
    if (!(!n.snap && !n.markline))
      switch (v) {
        case "drag-start":
          g();
          break;
        case "drag":
          return w();
        case "drag-end":
          l();
          break;
      }
  };
  return me(() => {
    s();
  }), {
    marklineEmit: M
  };
}
function pe(o = "x", n, e = "") {
  const t = document.createElement("div");
  return t.classList.add(`es-drager-markline-${o}`), t.style.position = "absolute", t.style.top = "0px", t.style.left = "0px", t.style.zIndex = "9999", t.style.backgroundColor = e, t.style.display = "none", o === "x" ? (t.style.height = "100%", t.style.width = "1px") : (t.style.height = "1px", t.style.width = "100%"), n.appendChild(t), t;
}
function He(o, n) {
  const e = { x: [], y: [] }, { width: t = 0, height: i = 0 } = n;
  return o.forEach((h) => {
    const {
      top: d,
      left: s,
      width: c,
      height: g
    } = h;
    e.y.push({ showTop: d, top: d }), e.y.push({ showTop: d, top: d - i }), e.y.push({
      showTop: d + g / 2,
      top: d + g / 2 - i / 2
    }), e.y.push({ showTop: d + g, top: d + g }), e.y.push({ showTop: d + g, top: d + g - i }), e.x.push({ showLeft: s, left: s }), e.x.push({ showLeft: s + c, left: s + c }), e.x.push({
      showLeft: s + c / 2,
      left: s + c / 2 - t / 2
    }), e.x.push({ showLeft: s + c, left: s + c - t }), e.x.push({ showLeft: s, left: s - t });
  }), e;
}
function We(o, n, e, {
  getBoundary: t,
  fixBoundary: i,
  checkDragerCollision: h,
  emit: d
}) {
  let s = 0, c = 0;
  return {
    onKeydown: (l) => {
      let { left: M, top: v } = n.value;
      if (s || (s = M), c || (c = v), ["ArrowRight", "ArrowLeft"].includes(l.key)) {
        const u = l.key === "ArrowRight";
        let r = u ? 1 : -1;
        o.snapToGrid && (r = u ? o.gridX : -o.gridX), M = M + r;
      } else if (["ArrowUp", "ArrowDown"].includes(l.key)) {
        const u = l.key === "ArrowDown";
        let r = u ? 1 : -1;
        o.snapToGrid && (r = u ? o.gridY : -o.gridY), v = v + r;
      }
      if (o.boundary) {
        const [u, r, f, p] = t();
        [M, v] = i(M, v, u, r, f, p);
      }
      n.value.left = M, n.value.top = v;
    },
    onKeyup: (l) => {
      ["ArrowRight", "ArrowLeft", "ArrowUp", "ArrowDown"].includes(l.key) && o.checkCollision && h() && (n.value.left = s, n.value.top = c), s = 0, c = 0;
    }
  };
}
function et(o, n, e) {
  const t = N(!1), i = N(!1), h = N({
    width: n.width,
    height: n.height,
    left: n.left,
    top: n.top,
    angle: n.angle
  }), { marklineEmit: d } = Ze(o, n), s = /* @__PURE__ */ new Set();
  function c(r) {
    if (s.add(r.button), n.disabled)
      return;
    t.value = !0, i.value = !0;
    let { clientX: f, clientY: p } = V(r);
    const { left: E, top: a } = h.value;
    let y = 0, m = 0, x = 0, b = 0;
    n.boundary && ([y, m, x, b] = g()), d("drag-start"), e && e("drag-start", h.value), te((B) => {
      if (s.size > 1)
        return;
      const { clientX: A, clientY: C } = V(B);
      let Y = (A - f) / n.scaleRatio + E, z = (C - p) / n.scaleRatio + a;
      if (n.snapToGrid) {
        let { left: D, top: T } = h.value;
        const Q = Y - D, Z = z - T;
        Y = D + J(Q, n.gridX), z = T + J(Z, n.gridY);
      }
      n.boundary && ([Y, z] = w(Y, z, y, m, x, b)), h.value.left = Y, h.value.top = z, e && e("drag", h.value), Ce(() => {
        const D = d("drag");
        n.snap && (D.diffX && (h.value.left += D.diffX), D.diffY && (h.value.top += D.diffY));
      });
    }, (B) => {
      n.checkCollision && l() && (h.value.top = a, h.value.left = E), s.clear(), t.value = !1, d("drag-end"), e && e("drag-end", h.value);
    });
  }
  const g = () => {
    let r = 0, f = 0;
    const { left: p, top: E, height: a, width: y, angle: m } = h.value, x = o.value.offsetParent || document.body, b = R(x, n.scaleRatio);
    if (m) {
      const A = R(o.value, n.scaleRatio);
      r = A.left - Math.floor(p - (A.width - y) + b.left), f = A.top - Math.floor(E - (A.height - a) + b.top);
    }
    const X = b.width - y, B = b.height - a;
    return [r, X - r, f, B - f, b.width, b.height];
  }, w = (r, f, p, E, a, y) => (r = r < p ? p : r, r = r > E ? E : r, f = f < a ? a : f, f = f > y ? y : f, [r, f]), l = () => {
    const r = o.value.offsetParent || document.body, f = Array.from(r.children).filter((p) => p !== o.value && p.classList.contains("es-drager"));
    for (let p = 0; p < f.length; p++) {
      const E = f[p];
      if (Qe(o.value, E, n.scaleRatio))
        return !0;
    }
  }, M = () => {
    i.value = !1;
  }, { onKeydown: v, onKeyup: u } = We(
    n,
    h,
    i,
    {
      getBoundary: g,
      fixBoundary: w,
      checkDragerCollision: l,
      emit: e
    }
  );
  return U(i, (r) => {
    r ? (e("focus", r), document.addEventListener("click", M, { once: !0 })) : e("blur", r), !n.disabledKeyEvent && (r ? (document.addEventListener("keydown", v), document.addEventListener("keyup", u)) : (document.removeEventListener("keydown", v), document.removeEventListener("keyup", u)));
  }), me(() => {
    if (o.value) {
      if (!h.value.width && !h.value.height) {
        const { width: r, height: f } = R(o.value, n.scaleRatio);
        h.value = {
          ...h.value,
          width: r || 100,
          height: f || 100
        };
      }
      o.value.addEventListener("mousedown", c), o.value.addEventListener("touchstart", c, {
        passive: !0
      });
    }
  }), Be(() => {
    document.removeEventListener("click", M), document.removeEventListener("keydown", v), document.removeEventListener("keyup", u);
  }), {
    isMousedown: t,
    selected: i,
    dragData: h,
    getBoundary: g,
    checkDragerCollision: l
  };
}
const tt = /* @__PURE__ */ j("div", { class: "es-drager-rotate-handle" }, [
  /* @__PURE__ */ j("svg", {
    viewBox: "0 0 1024 1024",
    xmlns: "http://www.w3.org/2000/svg"
  }, [
    /* @__PURE__ */ j("path", {
      fill: "currentColor",
      d: "M784.512 230.272v-50.56a32 32 0 1 1 64 0v149.056a32 32 0 0 1-32 32H667.52a32 32 0 1 1 0-64h92.992A320 320 0 1 0 524.8 833.152a320 320 0 0 0 320-320h64a384 384 0 0 1-384 384 384 384 0 0 1-384-384 384 384 0 0 1 643.712-282.88z"
    })
  ])
], -1), nt = /* @__PURE__ */ we({
  __name: "rotate",
  props: {
    modelValue: {
      type: Number,
      default: 0
    },
    element: {
      type: Object
    }
  },
  emits: [
    "update:modelValue",
    "rotate",
    "rotate-start",
    "rotate-end"
  ],
  setup(o, { emit: n }) {
    const e = o, t = n, i = N(null), h = K({
      get: () => e.modelValue,
      set: (s) => {
        t("update:modelValue", s);
      }
    });
    function d(s) {
      if (!e.element)
        return console.warn(
          "[es-drager] rotate component need drag element property"
        );
      s.stopPropagation();
      const { width: c, height: g, left: w, top: l } = e.element.getBoundingClientRect(), M = w + c / 2, v = l + g / 2;
      t("rotate-start", h.value), te(
        (u) => {
          const { clientX: r, clientY: f } = V(u), p = M - r, E = v - f, y = Math.atan2(E, p) * 180 / Math.PI - 90;
          h.value = (y + 360) % 360, t("rotate", h.value);
        },
        () => {
          t("rotate-end", h.value);
        }
      );
    }
    return (s, c) => ($(), W("div", {
      ref_key: "rotateRef",
      ref: i,
      class: "es-drager-rotate",
      onMousedown: d,
      onTouchstartPassive: d
    }, [
      O(s.$slots, "default", {}, () => [
        tt
      ])
    ], 544));
  }
});
const ot = ["data-side", "onMousedown", "onTouchstartPassive"], st = /* @__PURE__ */ j("div", { class: "es-drager-dot-handle" }, null, -1), be = /* @__PURE__ */ we({
  __name: "drager",
  props: $e,
  emits: [
    "change",
    "drag",
    "drag-start",
    "drag-end",
    "resize",
    "resize-start",
    "resize-end",
    "rotate",
    "rotate-start",
    "rotate-end",
    "focus",
    "blur"
  ],
  setup(o, { emit: n }) {
    const e = o, t = n, i = (a, ...y) => {
      t(a, ...y);
    }, h = N(null), { selected: d, dragData: s, isMousedown: c, getBoundary: g, checkDragerCollision: w } = et(
      h,
      e,
      i
    ), l = N(ve(0, e.resizeList)), M = K(() => e.resizable && !e.disabled), v = K(
      () => e.rotatable && !e.disabled && d.value
    ), u = K(() => {
      const { width: a, height: y, left: m, top: x, angle: b } = s.value, X = {};
      return a && (X.width = F(a)), y && (X.height = F(y)), {
        ...X,
        left: F(m),
        top: F(x),
        zIndex: e.zIndex,
        transform: `rotate(${b}deg)`,
        "--es-drager-color": e.color
      };
    });
    function r(a) {
      h.value || (h.value = a.$el || a);
    }
    function f(a) {
      l.value = ve(a, e.resizeList), i("rotate-end", s.value);
    }
    function p(a, y) {
      if (e.disabled)
        return;
      y.stopPropagation();
      const { clientX: m, clientY: x } = V(y), b = m, X = x, { width: B, height: A, left: C, top: Y } = s.value, z = C + B / 2, D = Y + A / 2, T = {
        width: B,
        height: A,
        centerX: z,
        centerY: D,
        rotateAngle: s.value.angle
      }, Q = a.side, { minWidth: Z, minHeight: ke, aspectRatio: oe, equalProportion: Le } = e;
      i("resize-start", s.value);
      let se = [];
      e.boundary && (se = g()), te((ie) => {
        const { clientX: Me, clientY: xe } = V(ie);
        let G = (Me - b) / e.scaleRatio, q = (xe - X) / e.scaleRatio;
        e.snapToGrid && (G = J(G, e.gridX), q = J(q, e.gridY));
        const Ae = Math.atan2(q, G), le = Ue(G, q), Ee = ie.shiftKey, re = Ae - ne(T.rotateAngle), Xe = le * Math.cos(re), Ye = le * Math.sin(re), ae = (Le || Ee) && !oe ? T.width / T.height : oe, {
          position: { centerX: ce, centerY: ue },
          size: { width: Te, height: ze }
        } = Oe(
          Q,
          { ...T, rotateAngle: T.rotateAngle },
          Xe,
          Ye,
          ae,
          Z,
          ke
        ), De = je({
          centerX: ce,
          centerY: ue,
          width: Te,
          height: ze,
          angle: s.value.angle
        });
        let _ = {
          ...s.value,
          ...Je(De, ce, ue)
        };
        e.maxWidth > 0 && (_.width = Math.min(_.width, e.maxWidth)), e.maxHeight > 0 && (_.height = Math.min(_.height, e.maxHeight)), e.boundary && (_ = E(_, se, ae)), s.value = _, i("resize", s.value);
      }, () => {
        e.checkCollision && w() && (s.value = { ...s.value, width: B, height: A, left: C, top: Y }), i("resize-end", s.value);
      });
    }
    function E(a, y, m) {
      const [x, b, X, B, A, C] = y, Y = a.left < x, z = a.left + a.width > A, D = a.top < X, T = a.top + a.height > C;
      return Y && (a.left = x, a.width = s.value.width), D && (a.top = X, a.height = s.value.height), (z || T) && (z && (a.left = s.value.left), T && (a.top = s.value.top), T || (a.width = A - a.left), z || (a.height = C - a.top)), (T || D) && m && (a.width = s.value.width, a.left = s.value.left), (z || Y) && m && (a.height = s.value.height, a.top = s.value.top), a;
    }
    return U(
      () => [e.width, e.height, e.left, e.top, e.angle],
      ([a, y, m, x, b], [X, B, A, C, Y]) => {
        a !== X && (s.value.width = a), y !== B && (s.value.height = y), m !== A && (s.value.left = m), x !== C && (s.value.top = x), b !== Y && (s.value.angle = b);
      }
    ), U(
      () => s.value,
      (a) => {
        t("change", { ...a });
      },
      { deep: !0 }
    ), U(
      () => e.selected,
      (a) => {
        d.value = a;
      },
      { immediate: !0 }
    ), (a, y) => ($(), he(Re(a.tag), {
      ref: r,
      class: _e([
        "es-drager",
        { disabled: a.disabled, dragging: S(c), selected: S(d), border: a.border }
      ]),
      style: de(u.value),
      onClick: y[3] || (y[3] = Ne(() => {
      }, ["stop"]))
    }, {
      default: fe(() => [
        O(a.$slots, "default"),
        M.value ? ($(!0), W(Se, { key: 0 }, Pe(l.value, (m, x) => ($(), W("div", {
          key: x,
          class: "es-drager-dot",
          "data-side": m.side,
          style: de({ ...m }),
          onMousedown: (b) => p(m, b),
          onTouchstartPassive: (b) => p(m, b)
        }, [
          O(a.$slots, "resize", Ie(Ke({ side: m.side })), () => [
            st
          ])
        ], 44, ot))), 128)) : ge("", !0),
        v.value ? ($(), he(nt, {
          key: 1,
          modelValue: S(s).angle,
          "onUpdate:modelValue": y[0] || (y[0] = (m) => S(s).angle = m),
          element: h.value,
          onRotate: y[1] || (y[1] = (m) => i("rotate", S(s))),
          onRotateStart: y[2] || (y[2] = (m) => i("rotate-start", S(s))),
          onRotateEnd: f
        }, {
          default: fe(() => [
            O(a.$slots, "rotate")
          ]),
          _: 3
        }, 8, ["modelValue", "element"])) : ge("", !0)
      ]),
      _: 3
    }, 8, ["class", "style"]));
  }
});
const it = (o) => {
  o.component("es-drager", be);
};
be.install = it;
export {
  $e as DragerProps,
  be as ESDrager,
  be as default,
  it as install
};
