// @opentiny/fluent-editor/lib/modules/ai/constants.cjs.js
"use strict";
Object.defineProperty(exports, Symbol.toStringTag, { value: "Module" });
const icons = require("./icons.cjs.js");
const INPUT_PLACEHOLDER = '请输入问题或"/"获取提示词';
const SELECT_PLACEHOLDER = "向我提问/选择操作";
const STOP_ANSWER = "停止回答";
const REPLACE_SELECT = "替换选中内容";
const INSERT_TEXT = "插入内容";
const INSERT_SUB_CONTENT_TEXT = "插入内容下方";
const REGENERATE = "重新生成";
const CLOSE = "关闭";
const THINK_TEXT = "正在为您分析并总结答案";
const RESULT_HEADER_TEXT = "根据您的诉求，已为您解答，具体如下：";
const MENU_TITLE_DATA = {
  editor: [
    { id: "1-1", text: "丰富内容", icon: icons.RICH_CONTENT_ICON },
    { id: "1-2", text: "精简内容", icon: icons.STREAMLINE_CONTENT_ICON },
    { id: "1-3", text: "修改标点符号", icon: icons.SYMBOL_ICON },
    { id: "1-4", text: "翻译", icon: icons.TRANSLATE_ICON }
  ],
  tone: [
    { id: "2-1", text: "更专业的" },
    { id: "2-2", text: "更直接的" },
    { id: "2-3", text: "更友善的" },
    { id: "2-4", text: "更口语化的" }
  ],
  adjust: [
    { id: "3-1", text: "提炼要点" },
    { id: "3-2", text: "归纳总结" },
    { id: "3-3", text: "转写成[代码块]" }
  ]
};
const MENU_ID_MAP = {
  editor: "subMenuEditorEl",
  tone: "subMenuToneEl",
  adjust: "subMenuAdjustEl"
};
exports.CLOSE = CLOSE;
exports.INPUT_PLACEHOLDER = INPUT_PLACEHOLDER;
exports.INSERT_SUB_CONTENT_TEXT = INSERT_SUB_CONTENT_TEXT;
exports.INSERT_TEXT = INSERT_TEXT;
exports.MENU_ID_MAP = MENU_ID_MAP;
exports.MENU_TITLE_DATA = MENU_TITLE_DATA;
exports.REGENERATE = REGENERATE;
exports.REPLACE_SELECT = REPLACE_SELECT;
exports.RESULT_HEADER_TEXT = RESULT_HEADER_TEXT;
exports.SELECT_PLACEHOLDER = SELECT_PLACEHOLDER;
exports.STOP_ANSWER = STOP_ANSWER;
exports.THINK_TEXT = THINK_TEXT;
