// @opentiny/fluent-editor/es/modules/mind-map/modules/context-menu.es.js
var __defProp = Object.defineProperty;
var __defNormalProp = (obj, key, value) => key in obj ? __defProp(obj, key, { enumerable: true, configurable: true, writable: true, value }) : obj[key] = value;
var __publicField = (obj, key, value) => __defNormalProp(obj, typeof key !== "symbol" ? key + "" : key, value);
import { I18N_LOCALE_CHANGE } from "quill-i18n";
import { registerMindMapI18N } from "../i18n/index.es.js";
class MindMapContextMenuHandler {
  constructor(quill, blot) {
    __publicField(this, "texts");
    __publicField(this, "lang");
    this.quill = quill;
    this.blot = blot;
    const i18nModule = this.quill.getModule("i18n");
    registerMindMapI18N(i18nModule);
    this.lang = i18nModule.getLocale();
    this.texts = this.resolveTexts();
    this.quill.on(I18N_LOCALE_CHANGE, (event) => {
      this.lang = event.locale;
      this.texts = this.resolveTexts();
      this.updateContextMenuItems();
    });
  }
  getText(key) {
    return this.texts[key];
  }
  resolveTexts() {
    return {
      copy: this.quill.getLangText("mindMap.contextMenu.copy"),
      cut: this.quill.getLangText("mindMap.contextMenu.cut"),
      paste: this.quill.getLangText("mindMap.contextMenu.paste"),
      deleteContent: this.quill.getLangText("mindMap.contextMenu.deleteContent"),
      deleteNode: this.quill.getLangText("mindMap.contextMenu.deleteNode"),
      deleteIcon: this.quill.getLangText("mindMap.contextMenu.deleteIcon")
    };
  }
  updateContextMenuItems() {
    if (!this.blot.contextMenu) return;
    const menuItems = this.blot.contextMenu.querySelectorAll(".ql-mind-map-context-menu-item");
    if (menuItems.length >= 4) {
      menuItems[0].textContent = this.texts.copy;
      menuItems[1].textContent = this.texts.cut;
      menuItems[2].textContent = this.texts.paste;
      menuItems[3].textContent = this.texts.delete;
    }
  }
}
const contextMenuHandlers = /* @__PURE__ */ new WeakMap();
function initContextMenu(blot, quill) {
  blot.contextMenu = document.createElement("div");
  blot.contextMenu.className = "ql-mind-map-context-menu";
  blot.domNode.appendChild(blot.contextMenu);
  const handler = new MindMapContextMenuHandler(quill, blot);
  contextMenuHandlers.set(blot, handler);
  addContextMenuItem(blot, handler.getText("copy"), () => handleCopy(blot));
  addContextMenuItem(blot, handler.getText("cut"), () => handleCut(blot));
  addContextMenuItem(blot, handler.getText("paste"), () => handlePaste(blot));
  addContextMenuItem(blot, handler.getText("delete"), () => handleDeleteContent(blot));
  if (blot.mindMap) {
    blot.mindMap.on("node_contextmenu", (e, node) => {
      e.preventDefault();
      e.stopPropagation();
      blot.currentNode = node;
      if (blot.contextMenu) {
        blot.contextMenu.style.display = "block";
        blot.contextMenu.style.left = `${e.clientX}px`;
        blot.contextMenu.style.top = `${e.clientY}px`;
      }
    });
  }
  const hideMenu = () => {
    if (blot.contextMenu) {
      blot.contextMenu.style.display = "none";
      blot.currentNode = null;
    }
  };
  blot.mindMap.on("node_click", hideMenu);
  blot.mindMap.on("draw_click", hideMenu);
  blot.mindMap.on("expand_btn_click", hideMenu);
  document.addEventListener("click", (e) => {
    if (blot.contextMenu && !blot.contextMenu.contains(e.target)) {
      hideMenu();
    }
  });
}
function addContextMenuItem(blot, text, onClick) {
  const item = document.createElement("div");
  item.className = "ql-mind-map-context-menu-item";
  item.textContent = text;
  item.style.padding = "5px 15px";
  item.style.cursor = "pointer";
  item.style.whiteSpace = "nowrap";
  item.addEventListener("click", onClick);
  item.addEventListener("mouseenter", () => {
    item.style.background = "#f5f5f5";
  });
  item.addEventListener("mouseleave", () => {
    item.style.background = "white";
  });
  blot.contextMenu.appendChild(item);
}
function handleCopy(blot) {
  blot.mindMap.renderer.copy();
  hideContextMenu(blot);
}
function handleCut(blot) {
  blot.mindMap.renderer.cut();
  hideContextMenu(blot);
}
function handlePaste(blot) {
  blot.mindMap.renderer.paste();
  hideContextMenu(blot);
}
function handleDeleteContent(blot) {
  if (blot.currentNode) {
    blot.currentNode.setText("");
    blot.data = blot.mindMap.getData({});
    blot.domNode.setAttribute("data-mind-map", JSON.stringify(blot.data));
  }
  hideContextMenu(blot);
}
function hideContextMenu(blot) {
  if (blot.contextMenu) {
    blot.contextMenu.style.display = "none";
  }
}
export {
  initContextMenu
};
