// @embedpdf/vue-pdf-viewer/dist/index.js
import { defineComponent, ref, onMounted, onBeforeUnmount, h } from "vue";
import EmbedPDF__default from "@embedpdf/snippet";
export * from "@embedpdf/snippet";
const PDFViewer = defineComponent({
  name: "PDFViewer",
  props: {
    /** Full configuration for the PDF viewer */
    config: {
      type: Object,
      default: () => ({})
    }
  },
  emits: {
    /** Emitted when the viewer is initialized */
    init: (_container) => true,
    /** Emitted when the registry is ready */
    ready: (_registry) => true
  },
  setup(props, { emit, expose, attrs }) {
    const containerRef = ref(null);
    const viewerRef = ref(null);
    expose({
      get container() {
        return viewerRef.value;
      },
      get registry() {
        var _a;
        return ((_a = viewerRef.value) == null ? void 0 : _a.registry) ?? null;
      }
    });
    onMounted(() => {
      if (!containerRef.value) return;
      const viewer = EmbedPDF__default.init({
        type: "container",
        target: containerRef.value,
        ...props.config
      });
      if (viewer) {
        viewerRef.value = viewer;
        emit("init", viewer);
        viewer.registry.then((registry) => {
          emit("ready", registry);
        });
      }
    });
    onBeforeUnmount(() => {
      if (viewerRef.value && containerRef.value) {
        containerRef.value.innerHTML = "";
        viewerRef.value = null;
      }
    });
    return () => h("div", {
      ref: containerRef,
      class: attrs.class,
      style: attrs.style
    });
  }
});
export {
  PDFViewer,
  PDFViewer as default
};
