// @logicflow/extension/es/pool/LaneModel.js
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (Object.prototype.hasOwnProperty.call(b, p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        if (typeof b !== "function" && b !== null)
            throw new TypeError("Class extends value " + String(b) + " is not a constructor or null");
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
var __read = (this && this.__read) || function (o, n) {
    var m = typeof Symbol === "function" && o[Symbol.iterator];
    if (!m) return o;
    var i = m.call(o), r, ar = [], e;
    try {
        while ((n === void 0 || n-- > 0) && !(r = i.next()).done) ar.push(r.value);
    }
    catch (error) { e = { error: error }; }
    finally {
        try {
            if (r && !r.done && (m = i["return"])) m.call(i);
        }
        finally { if (e) throw e.error; }
    }
    return ar;
};
var __spreadArray = (this && this.__spreadArray) || function (to, from, pack) {
    if (pack || arguments.length === 2) for (var i = 0, l = from.length, ar; i < l; i++) {
        if (ar || !(i in from)) {
            if (!ar) ar = Array.prototype.slice.call(from, 0, i);
            ar[i] = from[i];
        }
    }
    return to.concat(ar || Array.prototype.slice.call(from));
};
import { DynamicGroupNodeModel } from '../dynamic-group';
import { forEach } from 'lodash-es';
import { laneConfig } from './constant';
var LaneModel = /** @class */ (function (_super) {
    __extends(LaneModel, _super);
    function LaneModel() {
        var _this = _super.apply(this, __spreadArray([], __read(arguments), false)) || this;
        _this.isLane = true;
        _this.defaultZIndex = -1;
        return _this;
    }
    LaneModel.prototype.initNodeData = function (data) {
        var _a;
        _super.prototype.initNodeData.call(this, data);
        // 泳道特定配置
        this.width = data.width || laneConfig.defaultWidth;
        this.height = data.height || laneConfig.defaultHeight;
        this.draggable = true; // 允许拖拽（实际拖拽逻辑由泳池控制）
        this.resizable = true; // 允许调整大小
        this.rotatable = false; // 禁止旋转
        // 设置泳道层级
        // 如果传入了zIndex，使用传入的值，否则默认为2
        // 泳道层级应该比所属泳池高，确保显示在泳池上方
        this.defaultZIndex = data.zIndex || -1;
        this.setZIndex(this.defaultZIndex);
        this.autoToFront = true;
        this.text.editable = true;
        this.style.stroke = '#000';
        this.style.strokeWidth = 1;
        // 泳道属性配置
        this.properties = __assign(__assign({}, this.properties), { processRef: '', panels: ['processRef'], direction: ((_a = data.properties) === null || _a === void 0 ? void 0 : _a.direction) || 'vertical' });
        // 设置折叠尺寸（泳道不支持折叠，设置为与正常尺寸相同）
        this.collapsedWidth = this.width;
        this.collapsedHeight = this.height;
        this.expandWidth = this.width;
        this.expandHeight = this.height;
    };
    LaneModel.prototype.setAttributes = function () {
        _super.prototype.setAttributes.call(this);
        this.updateTextPosition();
    };
    LaneModel.prototype.setZIndex = function (zIndex) {
        this.zIndex = Math.min(zIndex, this.defaultZIndex) || this.defaultZIndex;
    };
    /**
     * 重写折叠方法 - 泳道不支持折叠
     */
    LaneModel.prototype.toggleCollapse = function () {
        // 泳道不支持折叠功能，保持展开状态
        this.isCollapsed = false;
        this.setProperties({ isCollapsed: false });
    };
    /**
     * 获取所属泳池ID
     */
    LaneModel.prototype.getPoolId = function () {
        var _this = this;
        try {
            // 检查graphModel是否存在
            if (!this.graphModel) {
                console.warn('GraphModel is not available');
                return null;
            }
            // 安全地获取泳池ID
            var poolModel = this.graphModel.nodes.find(function (node) {
                return node.children && node.children.has(_this.id);
            });
            return (poolModel === null || poolModel === void 0 ? void 0 : poolModel.id) || null;
        }
        catch (error) {
            console.error('Error getting pool ID:', error);
            return null;
        }
    };
    /**
     * 获取所属泳池模型
     */
    LaneModel.prototype.getPoolModel = function () {
        try {
            var poolId = this.getPoolId();
            if (!poolId) {
                return null;
            }
            // 检查graphModel是否存在
            if (!this.graphModel) {
                console.warn('GraphModel is not available for getting pool model');
                return null;
            }
            var poolModel = this.graphModel.getNodeModelById(poolId);
            return poolModel || null;
        }
        catch (error) {
            console.error('Error getting pool model:', error);
            return null;
        }
    };
    /**
     * 动态修改泳道属性
     */
    LaneModel.prototype.changeAttribute = function (_a) {
        var width = _a.width, height = _a.height, x = _a.x, y = _a.y;
        if (width)
            this.width = width; // 更新宽度
        if (height)
            this.height = height; // 更新高度
        if (x)
            this.x = x; // 更新X坐标
        if (y)
            this.y = y; // 更新Y坐标
    };
    /**
     * 重写获取数据方法，添加泳道特定属性
     */
    LaneModel.prototype.getData = function () {
        var data = _super.prototype.getData.call(this);
        // const poolModel = this.getPoolModel()
        return __assign(__assign({}, data), { properties: __assign(__assign({}, data.properties), { width: this.width, height: this.height, processRef: this.properties.processRef, direction: this.properties.direction }) });
    };
    /**
     * 重写 isAllowAppendIn，禁止 Lane 嵌套
     */
    LaneModel.prototype.isAllowAppendIn = function (nodeData) {
        // 禁止 Lane 节点被添加到 Lane 中
        return String(nodeData.type) !== 'lane';
    };
    /**
     * 获取需要移动的节点
     * @param groupModel
     */
    LaneModel.prototype.getNodesInGroup = function (groupModel) {
        var _this = this;
        var nodeIds = [];
        var parent = groupModel.properties.parent, isDragging = groupModel.isDragging;
        if (isDragging && parent) {
            nodeIds.push(parent);
        }
        forEach(Array.from(groupModel.children), function (nodeId) {
            var nodeModel = _this.graphModel.getNodeModelById(nodeId);
            // 只有非 Lane 类型的节点才会被带动
            if (nodeModel &&
                !nodeModel.isDragging &&
                String(nodeModel.type) !== 'lane') {
                nodeIds.push(nodeId);
            }
        });
        return nodeIds;
    };
    LaneModel.prototype.getNodeStyle = function () {
        var style = _super.prototype.getNodeStyle.call(this);
        style.strokeWidth = 2;
        return style;
    };
    /**
     * 获取文本样式
     */
    LaneModel.prototype.getTextStyle = function () {
        var _a = this.properties.isHorizontal, isHorizontal = _a === void 0 ? false : _a;
        var style = _super.prototype.getTextStyle.call(this);
        style.overflowMode = 'ellipsis';
        style.strokeWidth = 2;
        style.textWidth = isHorizontal ? this.height : this.width;
        style.textHeight = isHorizontal ? this.width : this.height;
        if (isHorizontal) {
            style.transform = 'rotate(-90deg)';
            style.textAlign = 'center';
        }
        return style;
    };
    /**
     * 获取子泳道
     */
    LaneModel.prototype.getSubNodes = function () {
        var _this = this;
        var children = [];
        Array.from(this.children).forEach(function (childId) {
            var childModel = _this.graphModel.getNodeModelById(childId);
            if (childModel) {
                children.push(childModel);
            }
        });
        return children;
    };
    /**
     * 初始化文本位置 - 根据布局方向设置文本位置
     */
    LaneModel.prototype.updateTextPosition = function () {
        if (this.properties.isHorizontal) {
            // 横向泳池：文本显示在左侧标题区域
            this.text.x = this.x - this.width / 2 + laneConfig.titleSize / 2;
            this.text.y = this.y;
        }
        else {
            // 纵向泳池：文本显示在顶部标题区域
            this.text.x = this.x;
            this.text.y = this.y - this.height / 2 + laneConfig.titleSize / 2;
        }
    };
    return LaneModel;
}(DynamicGroupNodeModel));
export { LaneModel };
export default null;
