// @opentiny/fluent-editor/lib/modules/index.cjs.js
"use strict";
Object.defineProperty(exports, Symbol.toStringTag, { value: "Module" });
require("./collaborative-editing/index.cjs.js");
require("./counter.cjs.js");
const customClipboard = require("./custom-clipboard.cjs.js");
require("./custom-image/index.cjs.js");
const customUploader = require("./custom-uploader.cjs.js");
const divider = require("./divider.cjs.js");
const emoji = require("./emoji.cjs.js");
require("./file/index.cjs.js");
const index = require("./flow-chart/index.cjs.js");
require("./link/index.cjs.js");
require("./mathlive/index.cjs.js");
require("./mention/index.cjs.js");
const index$1 = require("./mind-map/index.cjs.js");
const index$2 = require("./shortcut-key/index.cjs.js");
require("./syntax.cjs.js");
const index$3 = require("./table-up/index.cjs.js");
require("./toolbar/index.cjs.js");
const I18n = require("quill-i18n");
const QuillShortcutKey = require("quill-shortcut-key");
exports.CustomClipboard = customClipboard.CustomClipboard;
exports.FileUploader = customUploader.FileUploader;
exports.DividerBlot = divider.DividerBlot;
exports.EmojiModule = emoji.EmojiModule;
exports.FlowChartModule = index.FlowChartModule;
exports.MindMapModule = index$1.MindMapModule;
exports.ShortCutKey = index$2.ShortCutKey;
exports.shortKey = index$2.shortKey;
exports.generateTableUp = index$3.generateTableUp;
Object.keys(I18n).forEach((k) => {
  if (k !== "default" && !Object.prototype.hasOwnProperty.call(exports, k)) Object.defineProperty(exports, k, {
    enumerable: true,
    get: () => I18n[k]
  });
});
Object.keys(QuillShortcutKey).forEach((k) => {
  if (k !== "default" && !Object.prototype.hasOwnProperty.call(exports, k)) Object.defineProperty(exports, k, {
    enumerable: true,
    get: () => QuillShortcutKey[k]
  });
});
