// @opentiny/fluent-editor/lib/modules/flow-chart/modules/control-panel.cjs.js
"use strict";
var __defProp = Object.defineProperty;
var __defNormalProp = (obj, key, value) => key in obj ? __defProp(obj, key, { enumerable: true, configurable: true, writable: true, value }) : obj[key] = value;
var __publicField = (obj, key, value) => __defNormalProp(obj, typeof key !== "symbol" ? key + "" : key, value);
Object.defineProperty(exports, Symbol.toStringTag, { value: "Module" });
const index = require("../i18n/index.cjs.js");
const icons = require("../icons.cjs.js");
class FlowChartControlPanelHandler {
  constructor(quill, blot) {
    __publicField(this, "texts");
    __publicField(this, "lang");
    this.quill = quill;
    this.blot = blot;
    const i18nModule = this.quill.getModule("i18n");
    index.registerFlowChartI18N(i18nModule);
    this.lang = i18nModule.getLocale();
    this.texts = this.resolveTexts();
    this.quill.on("i18n-locale-change", (event) => {
      this.lang = event.locale;
      this.texts = this.resolveTexts();
      this.updateControlPanelTexts();
    });
  }
  getText(key) {
    return this.texts[key];
  }
  resolveTexts() {
    return {
      exportTitle: this.quill.getLangText("flowChart.controlPanel.exportTitle"),
      zoomOutTitle: this.quill.getLangText("flowChart.controlPanel.zoomOutTitle"),
      zoomInTitle: this.quill.getLangText("flowChart.controlPanel.zoomInTitle"),
      fitTitle: this.quill.getLangText("flowChart.controlPanel.fitTitle"),
      backTitle: this.quill.getLangText("flowChart.controlPanel.backTitle"),
      forwardTitle: this.quill.getLangText("flowChart.controlPanel.forwardTitle"),
      setEdgeTypeTitle: this.quill.getLangText("flowChart.controlPanel.setEdgeTypeTitle"),
      panelStatusTitle: this.quill.getLangText("flowChart.controlPanel.panelStatusTitle"),
      screenTypeTitle: this.quill.getLangText("flowChart.controlPanel.screenTypeTitle")
    };
  }
  updateControlPanelTexts() {
    const controlItems = this.blot.domNode.querySelectorAll(".ql-flow-chart-control-item");
    const controlItemMap = {
      "zoom-out": "zoomOutTitle",
      "zoom-in": "zoomInTitle",
      "fit": "fitTitle",
      "back": "backTitle",
      "forward": "forwardTitle",
      "set-edge-type": "setEdgeTypeTitle",
      "panel-status": "panelStatusTitle",
      "screen-type": "screenTypeTitle"
    };
    controlItems.forEach((item) => {
      const controlType = item.dataset.controlType;
      if (controlType && controlItemMap[controlType] && this.texts[controlItemMap[controlType]]) {
        item.title = this.texts[controlItemMap[controlType]];
      }
    });
  }
}
const controlPanelHandlers = /* @__PURE__ */ new WeakMap();
const DISABLED_OPACITY = "0.5";
const ENABLED_OPACITY = "1";
function createControlPanel(blot, quill) {
  const controlPanel = document.createElement("div");
  controlPanel.className = "ql-flow-chart-control";
  const controlRightUpPanel = document.createElement("div");
  controlRightUpPanel.className = "ql-flow-chart-right-up-control";
  const handler = new FlowChartControlPanelHandler(quill, blot);
  controlPanelHandlers.set(blot, handler);
  const zoomOutBtn = createControlItem("zoom-out", handler.getText("zoomOutTitle"), () => handleZoomOut(blot));
  const zoomInBtn = createControlItem("zoom-in", handler.getText("zoomInTitle"), () => handleZoomIn(blot));
  const resetBtn = createControlItem("fit", handler.getText("fitTitle"), () => handleResetZoom(blot));
  const backBtn = createControlItem("back", handler.getText("backTitle"), () => handleUndo(blot));
  const forwardBtn = createControlItem("forward", handler.getText("forwardTitle"), () => handleRedo(blot));
  const setEdgeTypeBtn = createControlItem("set-edge-type", handler.getText("setEdgeTypeTitle"), () => handleSetEdgeType(blot));
  const panelStatusBtn = createControlItem("panel-status", handler.getText("panelStatusTitle"));
  const screenTypeBtn = createControlItem("screen-type", handler.getText("screenTypeTitle"), () => handleScreenTypeBtn(blot));
  const updateButtonState = (historyData) => {
    if (!historyData.data) {
      backBtn.style.opacity = DISABLED_OPACITY;
      backBtn.style.cursor = "not-allowed";
      forwardBtn.style.opacity = DISABLED_OPACITY;
      forwardBtn.style.cursor = "not-allowed";
      return;
    }
    const isUndoAvailable = historyData.data.undoAble || historyData.data.undos.length < 0;
    const isRedoAvailable = historyData.data.redoAble || historyData.data.redos.length > 0;
    if (backBtn) {
      backBtn.style.opacity = isUndoAvailable ? ENABLED_OPACITY : DISABLED_OPACITY;
      backBtn.style.cursor = isUndoAvailable ? "pointer" : "not-allowed";
    }
    if (forwardBtn) {
      forwardBtn.style.opacity = isRedoAvailable ? ENABLED_OPACITY : DISABLED_OPACITY;
      forwardBtn.style.cursor = isRedoAvailable ? "pointer" : "not-allowed";
    }
  };
  updateButtonState(blot.flowChart.history);
  blot.flowChart.on("history:change", (data) => {
    updateButtonState(data);
  });
  setTimeout(() => {
    const controlLeftUpPanel = blot.domNode.querySelector(".lf-dndpanel");
    controlLeftUpPanel.append(setEdgeTypeBtn);
  }, 0);
  controlRightUpPanel.append(panelStatusBtn);
  blot.domNode.appendChild(controlRightUpPanel);
  controlPanel.append(zoomOutBtn, zoomInBtn, resetBtn, screenTypeBtn, backBtn, forwardBtn);
  blot.domNode.appendChild(controlPanel);
}
function handleUndo(blot) {
  if (blot.flowChart) {
    blot.flowChart.undo();
  }
}
function handleRedo(blot) {
  if (blot.flowChart) {
    blot.flowChart.redo();
  }
}
function handleZoomIn(blot) {
  if (blot.flowChart) {
    blot.flowChart.zoom(true);
  }
}
function handleZoomOut(blot) {
  if (blot.flowChart) {
    blot.flowChart.zoom(false);
  }
}
function handleResetZoom(blot) {
  if (blot.flowChart) {
    blot.flowChart.resetZoom();
  }
}
function createControlItem(iconClass, title, onClick, disabled = false) {
  const controlItem = document.createElement("div");
  controlItem.className = "ql-flow-chart-control-item";
  controlItem.title = title;
  controlItem.dataset.controlType = iconClass;
  controlItem.style.cursor = disabled ? "not-allowed" : "pointer";
  const iconMap = {
    "back": icons.backIcon,
    "forward": icons.forwardIcon,
    "zoom-out": icons.zoomOutIcon,
    "zoom-in": icons.zoomInIcon,
    "fit": icons.fitIcon,
    "screen-type": icons.screenTypeIcon,
    "panel-status": icons.contractIcon,
    "set-edge-type": icons.lineIcon
  };
  const icon = document.createElement("i");
  icon.innerHTML = iconMap[iconClass] || "";
  controlItem.appendChild(icon);
  if (!disabled) {
    controlItem.addEventListener("click", onClick);
  }
  return controlItem;
}
function handleSetEdgeType(blot) {
  const controlLeftUpPanel = blot.domNode.querySelector(".lf-dndpanel");
  if (!controlLeftUpPanel) return;
  let edgeTypePanel = controlLeftUpPanel.querySelector(".ql-flow-chart-edge-panel");
  if (!edgeTypePanel) {
    edgeTypePanel = document.createElement("div");
    edgeTypePanel.className = "ql-flow-chart-edge-panel";
    edgeTypePanel.style.display = "flex";
    edgeTypePanel.style.justifyContent = "space-around";
    edgeTypePanel.style.flexWrap = "nowrap";
    edgeTypePanel.style.width = "auto";
    edgeTypePanel.style.minWidth = "100%";
    edgeTypePanel.style.padding = "8px";
    edgeTypePanel.style.boxSizing = "border-box";
    edgeTypePanel.style.position = "absolute";
    edgeTypePanel.style.top = "170px";
    controlLeftUpPanel.appendChild(edgeTypePanel);
    const edgeTypes = [
      {
        name: "line",
        displayName: "直线",
        icon: icons.lineIcon
      },
      {
        name: "polyline",
        displayName: "折线",
        icon: icons.polyLineIcon
      },
      {
        name: "bezier",
        displayName: "贝塞尔曲线",
        icon: icons.bezierIcon
      }
    ];
    edgeTypes.forEach((edgeType) => {
      const edgeItem = document.createElement("div");
      edgeItem.className = "ql-flow-chart-edge-item";
      edgeItem.style.padding = "8px";
      edgeItem.style.cursor = "pointer";
      edgeItem.style.textAlign = "center";
      const edgeIcon = document.createElement("div");
      edgeIcon.className = `ql-flow-chart-edge-type-icon`;
      edgeIcon.style.width = "24px";
      edgeIcon.style.height = "24px";
      edgeIcon.style.margin = "0 auto";
      if (edgeType.icon) {
        edgeIcon.innerHTML = edgeType.icon;
      }
      edgeItem.appendChild(edgeIcon);
      edgeItem.addEventListener("click", () => {
        const { edges = [] } = blot.flowChart.getSelectElements();
        if (edges.length > 0) {
          edges.forEach((edge) => {
            blot.flowChart.changeEdgeType(edge.id, edgeType.name);
          });
          blot.data = blot.flowChart.getGraphData();
          blot.domNode.setAttribute("data-flow-chart", JSON.stringify(blot.data));
        }
        edgeTypePanel.style.display = "none";
      });
      edgeTypePanel.appendChild(edgeItem);
    });
  } else {
    edgeTypePanel.style.display = "flex";
    edgeTypePanel.style.justifyContent = "space-around";
    edgeTypePanel.style.flexWrap = "nowrap";
  }
  const handleOutsideClick = (e) => {
    let setEdgeTypeBtn = null;
    const controlItems = controlLeftUpPanel.querySelectorAll(".ql-flow-chart-control-item");
    controlItems.forEach((item) => {
      const iconEl = item.querySelector("i");
      if (iconEl && iconEl.className.includes("set-edge-type")) {
        setEdgeTypeBtn = item;
      }
    });
    if (!edgeTypePanel.contains(e.target) && (!setEdgeTypeBtn || !setEdgeTypeBtn.contains(e.target))) {
      edgeTypePanel.style.display = "none";
      document.removeEventListener("click", handleOutsideClick);
    }
  };
  document.removeEventListener("click", handleOutsideClick);
  setTimeout(() => {
    document.addEventListener("click", handleOutsideClick);
  }, 0);
}
function handleScreenTypeBtn(blot) {
  const screenTypeBtn = blot.domNode.querySelector('[data-control-type="screen-type"]');
  if (!screenTypeBtn || !blot.domNode) return;
  const flowChartContainer = blot.domNode;
  const isFullscreen = flowChartContainer.style.position === "fixed";
  if (isFullscreen) {
    const originalPosition = flowChartContainer.getAttribute("data-original-position");
    const originalWidth = flowChartContainer.getAttribute("data-original-width");
    const originalHeight = flowChartContainer.getAttribute("data-original-height");
    if (originalWidth && originalHeight) {
      flowChartContainer.style.position = originalPosition;
      flowChartContainer.style.width = originalWidth;
      flowChartContainer.style.height = originalHeight;
      flowChartContainer.style.zIndex = "0";
    }
    const iconElement = screenTypeBtn.querySelector("i");
    if (iconElement) {
      iconElement.innerHTML = icons.screenTypeIcon;
    }
  } else {
    flowChartContainer.setAttribute("data-original-position", flowChartContainer.style.position || "");
    flowChartContainer.setAttribute("data-original-width", flowChartContainer.style.width || "");
    flowChartContainer.setAttribute("data-original-height", flowChartContainer.style.height || "");
    flowChartContainer.style.position = "fixed";
    flowChartContainer.style.top = "0";
    flowChartContainer.style.left = "0";
    flowChartContainer.style.width = "100vw";
    flowChartContainer.style.height = "100vh";
    flowChartContainer.style.zIndex = "100";
    const iconElement = screenTypeBtn.querySelector("i");
    if (iconElement) {
      iconElement.innerHTML = icons.screenReduceIcon;
    }
  }
  blot.flowChart.resize();
  blot.flowChart.translateCenter();
}
exports.createControlPanel = createControlPanel;
