// @opentiny/fluent-editor/lib/modules/custom-image/specs/blot-spec.cjs.js
"use strict";
var __defProp = Object.defineProperty;
var __defNormalProp = (obj, key, value) => key in obj ? __defProp(obj, key, { enumerable: true, configurable: true, writable: true, value }) : obj[key] = value;
var __publicField = (obj, key, value) => __defNormalProp(obj, typeof key !== "symbol" ? key + "" : key, value);
Object.defineProperty(exports, Symbol.toStringTag, { value: "Module" });
require("../actions/index.cjs.js");
const toolbarAction = require("../actions/toolbar-action.cjs.js");
const customResizeAction = require("../actions/custom-resize-action.cjs.js");
const deleteAction = require("../actions/delete-action.cjs.js");
class BlotSpec {
  constructor(formatter) {
    __publicField(this, "formatter");
    this.formatter = formatter;
  }
  init() {
  }
  getActions() {
    return [toolbarAction.ImageToolbarAction, customResizeAction.CustomResizeAction, deleteAction.DeleteAction];
  }
  getTargetElement() {
    return null;
  }
  getOverlayElement() {
    return this.getTargetElement();
  }
  setSelection() {
    this.formatter.quill.setSelection(null);
  }
  onHide() {
  }
}
exports.BlotSpec = BlotSpec;
