// @tiptap/extension-invisible-characters/dist/index.cjs
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);

// src/index.ts
var index_exports = {};
__export(index_exports, {
  HardBreakNode: () => HardBreakNode,
  InvisibleCharacter: () => InvisibleCharacter,
  InvisibleCharacters: () => InvisibleCharacters,
  InvisibleNode: () => InvisibleNode,
  ParagraphNode: () => ParagraphNode,
  SpaceCharacter: () => SpaceCharacter,
  default: () => index_default
});
module.exports = __toCommonJS(index_exports);

// src/invisible-characters.ts
var import_core = require("@tiptap/core");

// src/plugin/InvisibleCharactersPlugin.ts
var import_state = require("@tiptap/pm/state");
var import_view = require("@tiptap/pm/view");

// src/plugin/reducers.ts
var stateReducer = (state, visible) => {
  if (visible === void 0) {
    return state;
  }
  return { ...state, visible };
};

// src/plugin/style.ts
var style = `.tiptap-invisible-character {
  height: 0;
  padding: 0;
  pointer-events: none;
  user-select: none;
  width: 0;
}

.tiptap-invisible-character::before {
  caret-color: inherit;
  color: #aaa;
  display: inline-block;
  font-style: normal;
  font-weight: 400;
  line-height: 1em;
  width: 0;
}

.tiptap-invisible-character--space::before {
  content: '\xB7'
}

.tiptap-invisible-character--break::before {
  content: '\xAC'
}

.tiptap-invisible-character--paragraph::before {
  content: '\xB6'
}

.tiptap-invisible-character + img.ProseMirror-separator {
  height: 0 !important;
  pointer-events: none;
  user-select: none;
  width: 0 !important;
}

.is-empty[data-placeholder].has-focus > .tiptap-invisible-character {
  display: none;
}
`;

// src/plugin/utils/create-style-tag.ts
var createStyleTag = (style2, nonce) => {
  const tiptapStyleTag = document.querySelector("style[data-tiptap-extension-invisible-characters-style]");
  if (tiptapStyleTag !== null) {
    return tiptapStyleTag;
  }
  const styleNode = document.createElement("style");
  if (nonce) {
    styleNode.setAttribute("nonce", nonce);
  }
  styleNode.setAttribute("data-tiptap-extension-invisible-characters-style", "");
  styleNode.innerHTML = style2;
  document.getElementsByTagName("head")[0].appendChild(styleNode);
  return styleNode;
};

// src/plugin/utils/get-updated-ranges.ts
var getUpdatedRanges = ({ mapping }) => {
  const ranges = [];
  mapping.maps.forEach((stepMap, i) => {
    stepMap.forEach((_oldStart, _oldEnd, newStart, newEnd) => {
      ranges.push([mapping.slice(i + 1).map(newStart), mapping.slice(i + 1).map(newEnd)]);
    });
  });
  return ranges;
};

// src/plugin/InvisibleCharactersPlugin.ts
var InvisibleCharactersPluginKey = new import_state.PluginKey("invisibleCharacters");
var InvisibleCharactersPlugin = (state, options) => {
  const key = InvisibleCharactersPluginKey;
  const emptyDecorationSet = import_view.DecorationSet.create(state.doc, []);
  const addDecorationsBetweenPositions = (from, to, editorState, decorations) => {
    return options.builders.sort((builderA, builderB) => {
      if (builderA.priority > builderB.priority) {
        return 1;
      }
      return -1;
    }).reduce((newDecos, builder) => {
      return builder.createDecoration(from, to, editorState.doc, newDecos);
    }, decorations);
  };
  return new import_state.Plugin({
    key,
    state: {
      init: () => {
        const { $from, $to } = new import_state.AllSelection(state.doc);
        if (options.injectCSS && document) {
          createStyleTag(style, options.injectNonce);
        }
        return {
          visible: options.visible,
          decorations: addDecorationsBetweenPositions($from.pos, $to.pos, state, import_view.DecorationSet.empty)
        };
      },
      apply: (tr, pluginState, _, currentState) => {
        const newPluginState = stateReducer(pluginState, tr.getMeta("setInvisibleCharactersVisible"));
        const decorations = getUpdatedRanges(tr).reduce(
          (nextDecorations, [from, to]) => {
            return addDecorationsBetweenPositions(from, to, currentState, nextDecorations);
          },
          newPluginState.decorations.map(tr.mapping, tr.doc)
        );
        return {
          ...newPluginState,
          decorations
        };
      }
    },
    props: {
      decorations(editorState) {
        const pluginState = this.getState(editorState);
        const visible = pluginState == null ? void 0 : pluginState.visible;
        const decorations = pluginState == null ? void 0 : pluginState.decorations;
        return visible ? decorations : emptyDecorationSet;
      }
    }
  });
};

// src/plugin/utils/create-decoration-widget.ts
var import_view2 = require("@tiptap/pm/view");
var createDecorationWidget = (pos, type, content) => {
  const createElement = () => {
    const newElement = document.createElement("span");
    newElement.classList.add("tiptap-invisible-character");
    newElement.classList.add(`tiptap-invisible-character--${type}`);
    if (content) {
      newElement.textContent = content;
    }
    return newElement;
  };
  return import_view2.Decoration.widget(pos, createElement, {
    key: type,
    marks: [],
    side: 1e3
  });
};

// src/plugin/InvisibleNode.ts
var defaultPosition = (node, pos) => pos + node.nodeSize - 1;
var InvisibleNode = class {
  constructor(options) {
    this.predicate = options.predicate;
    this.type = options.type;
    this.position = options.position || defaultPosition;
    this.content = options.content;
    this.priority = options.priority || 100;
  }
  createDecoration(from, to, doc, decorations) {
    let newDecorations = decorations;
    doc.nodesBetween(from, to, (node, pos) => {
      if (this.test(node)) {
        const decorationPosition = this.position(node, pos);
        const oldDecorations = newDecorations.find(
          decorationPosition,
          decorationPosition,
          (spec) => spec.key === this.type
        );
        newDecorations = newDecorations.remove(oldDecorations).add(doc, [createDecorationWidget(decorationPosition, this.type, this.content)]);
      }
    });
    return newDecorations;
  }
  test(value) {
    return this.predicate(value);
  }
};

// src/plugin/invisible-characters/hardBreak.ts
var HardBreakNode = class extends InvisibleNode {
  constructor() {
    super({
      type: "break",
      predicate: (node) => node.type === node.type.schema.nodes.hardBreak
    });
  }
};

// src/plugin/invisible-characters/paragraph.ts
var ParagraphNode = class extends InvisibleNode {
  constructor() {
    super({
      type: "paragraph",
      predicate: (node) => node.type === node.type.schema.nodes.paragraph
    });
  }
};

// src/plugin/utils/text-between.ts
var textBetween = (from, to, doc) => {
  const positions = [];
  doc.nodesBetween(from, to, (node, pos) => {
    var _a;
    if (node.isText) {
      const offset = Math.max(from, pos) - pos;
      positions.push({
        pos: pos + offset,
        text: ((_a = node.text) == null ? void 0 : _a.slice(offset, to - pos)) || ""
      });
    }
  });
  return positions;
};

// src/plugin/InvisibleCharacter.ts
var InvisibleCharacter = class {
  constructor(options) {
    this.predicate = options.predicate;
    this.type = options.type;
    this.content = options.content;
    this.priority = options.priority || 100;
  }
  createDecoration(from, to, doc, decorations) {
    const textContent = textBetween(from, to, doc);
    return textContent.reduce((oldDecorations, currentPosition) => {
      return currentPosition.text.split("").reduce((innerDecorations, char, i) => {
        return this.test(char) ? innerDecorations.add(doc, [createDecorationWidget(currentPosition.pos + i, this.type, this.content)]) : innerDecorations;
      }, oldDecorations);
    }, decorations);
  }
  test(value) {
    return this.predicate(value);
  }
};

// src/plugin/invisible-characters/space.ts
var SpaceCharacter = class extends InvisibleCharacter {
  constructor() {
    super({
      type: "space",
      predicate: (char) => char === " "
    });
  }
};

// src/invisible-characters.ts
var InvisibleCharacters = import_core.Extension.create({
  name: "invisibleCharacters",
  addOptions() {
    return {
      visible: true,
      builders: [new SpaceCharacter(), new ParagraphNode(), new HardBreakNode()],
      injectCSS: true,
      injectNonce: void 0
    };
  },
  addProseMirrorPlugins() {
    return [InvisibleCharactersPlugin(this.editor.state, this.options)];
  },
  addStorage() {
    return {
      visibility: () => this.options.visible
    };
  },
  onBeforeCreate() {
    this.storage.visibility = () => {
      var _a;
      return (_a = InvisibleCharactersPluginKey.getState(this.editor.state)) == null ? void 0 : _a.visible;
    };
  },
  addCommands() {
    return {
      showInvisibleCharacters: (visibility = true) => ({ dispatch, tr }) => {
        if (dispatch) {
          tr.setMeta("setInvisibleCharactersVisible", visibility);
        }
        return true;
      },
      hideInvisibleCharacters: () => ({ dispatch, tr }) => {
        if (dispatch) {
          tr.setMeta("setInvisibleCharactersVisible", false);
        }
        return true;
      },
      toggleInvisibleCharacters: () => ({ dispatch, tr, state }) => {
        var _a;
        const visibility = !((_a = InvisibleCharactersPluginKey.getState(state)) == null ? void 0 : _a.visible);
        if (dispatch) {
          tr.setMeta("setInvisibleCharactersVisible", visibility);
        }
        return true;
      }
    };
  }
});

// src/index.ts
var index_default = InvisibleCharacters;
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  HardBreakNode,
  InvisibleCharacter,
  InvisibleCharacters,
  InvisibleNode,
  ParagraphNode,
  SpaceCharacter
});
