// @opentiny/fluent-editor/lib/modules/collaborative-editing/module.cjs.js
"use strict";
var __defProp = Object.defineProperty;
var __defNormalProp = (obj, key, value) => key in obj ? __defProp(obj, key, { enumerable: true, configurable: true, writable: true, value }) : obj[key] = value;
var __publicField = (obj, key, value) => __defNormalProp(obj, typeof key !== "symbol" ? key + "" : key, value);
Object.defineProperty(exports, Symbol.toStringTag, { value: "Module" });
const collaborativeEditing = require("./collaborative-editing.cjs.js");
class CollaborationModule {
  constructor(quill, options) {
    __publicField(this, "collaborativeEditor");
    this.quill = quill;
    this.options = options;
    this.collaborativeEditor = new collaborativeEditing.CollaborativeEditor(quill, options);
    window.addEventListener(
      "beforeunload",
      () => {
        void this.collaborativeEditor.destroy().catch((err) => console.warn("[yjs] destroy failed:", err));
      },
      { once: true }
    );
  }
  getCursors() {
    return this.collaborativeEditor.getCursors();
  }
  getAwareness() {
    return this.collaborativeEditor.getAwareness();
  }
  getProvider() {
    return this.collaborativeEditor.getProvider();
  }
  async destroy() {
    await this.collaborativeEditor.destroy();
  }
}
exports.CollaborationModule = CollaborationModule;
