// @opentiny/fluent-editor/lib/modules/file/modules/file-module.cjs.js
"use strict";
var __defProp = Object.defineProperty;
var __defNormalProp = (obj, key, value) => key in obj ? __defProp(obj, key, { enumerable: true, configurable: true, writable: true, value }) : obj[key] = value;
var __publicField = (obj, key, value) => __defNormalProp(obj, typeof key !== "symbol" ? key + "" : key, value);
Object.defineProperty(exports, Symbol.toStringTag, { value: "Module" });
const Quill = require("quill");
const file = require("../formats/file.cjs.js");
const fileBar = require("./file-bar.cjs.js");
class FileModule {
  constructor(quill) {
    __publicField(this, "fileBar");
    this.quill = quill;
    this.quill = quill;
    quill.root.addEventListener("click", (event) => this.clickEvent(event), false);
  }
  static register() {
    Quill.register("formats/file", file.File, true);
  }
  clickEvent(event) {
    const target = event.target;
    const fileDom = target.closest("a.ql-file-item");
    const fileBar$1 = target.closest(".ql-file-bar");
    if (fileDom) {
      event.preventDefault();
      if (!this.quill.isEnabled()) {
        this.downloadFile(fileDom);
        return;
      }
      if (this.fileBar) {
        this.fileBar.destroy();
      }
      this.fileBar = new fileBar.FileBar(this.quill, fileDom);
    } else if (this.fileBar && !fileBar$1) {
      event.preventDefault();
      this.fileBar.destroy();
      this.fileBar = null;
    }
  }
  downloadFile(fileDom) {
    var _a;
    const fileName = fileDom.dataset.title || "";
    const fileDownloadUrl = fileDom.getAttribute("href") || "";
    if (fileDownloadUrl) {
      const a = document.createElement("a");
      a.href = fileDownloadUrl;
      a.target = "_blank";
      a.id = "exppub";
      a.download = fileName;
      document.body.appendChild(a);
      const alink = document.getElementById("exppub");
      alink == null ? void 0 : alink.click();
      (_a = alink == null ? void 0 : alink.parentNode) == null ? void 0 : _a.removeChild(a);
    }
  }
}
exports.FileModule = FileModule;
