// @opentiny/fluent-editor/lib/index.cjs.js
"use strict";
Object.defineProperties(exports, { __esModule: { value: true }, [Symbol.toStringTag]: { value: "Module" } });
require("./fluent-editor.cjs.js");
require("./attributors/index.cjs.js");
const index = require("./config/index.cjs.js");
require("./formats/index.cjs.js");
require("./modules/index.cjs.js");
const QuillShortcutKey = require("quill-shortcut-key");
const core = require("quill/core");
const fluentEditor = require("./core/fluent-editor.cjs.js");
const fontSize = require("./attributors/font-size.cjs.js");
const fontStyle = require("./attributors/font-style.cjs.js");
const lineHeight = require("./attributors/line-height.cjs.js");
const textIndent = require("./attributors/text-indent.cjs.js");
const editor_config = require("./config/editor.config.cjs.js");
const editor_utils = require("./config/editor.utils.cjs.js");
const emoji = require("./formats/emoji.cjs.js");
const softBreak = require("./formats/soft-break.cjs.js");
const strike = require("./formats/strike.cjs.js");
const video = require("./formats/video.cjs.js");
const module$1 = require("./modules/collaborative-editing/module.cjs.js");
const customClipboard = require("./modules/custom-clipboard.cjs.js");
const action = require("./modules/custom-image/actions/action.cjs.js");
const customResizeAction = require("./modules/custom-image/actions/custom-resize-action.cjs.js");
const deleteAction = require("./modules/custom-image/actions/delete-action.cjs.js");
const imageToolbarButtons = require("./modules/custom-image/actions/image-toolbar-buttons.cjs.js");
const toolbar = require("./modules/custom-image/actions/toolbar.cjs.js");
const toolbarAction = require("./modules/custom-image/actions/toolbar-action.cjs.js");
const blotFormatter = require("./modules/custom-image/blot-formatter.cjs.js");
const image = require("./modules/custom-image/image.cjs.js");
const previewModal = require("./modules/custom-image/preview/preview-modal.cjs.js");
const blotSpec = require("./modules/custom-image/specs/blot-spec.cjs.js");
const customImageSpec = require("./modules/custom-image/specs/custom-image-spec.cjs.js");
const imageSpec = require("./modules/custom-image/specs/image-spec.cjs.js");
const customUploader = require("./modules/custom-uploader.cjs.js");
const divider = require("./modules/divider.cjs.js");
const emoji$1 = require("./modules/emoji.cjs.js");
const file = require("./modules/file/formats/file.cjs.js");
const fileBar = require("./modules/file/modules/file-bar.cjs.js");
const fileModule = require("./modules/file/modules/file-module.cjs.js");
const index$1 = require("./modules/flow-chart/index.cjs.js");
const link = require("./modules/link/formats/link.cjs.js");
const tooltip = require("./modules/link/modules/tooltip.cjs.js");
const formats = require("./modules/mathlive/formats.cjs.js");
const module$2 = require("./modules/mathlive/module.cjs.js");
const tooltip$1 = require("./modules/mathlive/tooltip.cjs.js");
const mention = require("./modules/mention/mention.cjs.js");
const mentionLink = require("./modules/mention/mention-link.cjs.js");
const index$2 = require("./modules/mind-map/index.cjs.js");
const index$3 = require("./modules/shortcut-key/index.cjs.js");
const index$4 = require("./modules/table-up/index.cjs.js");
const betterPicker = require("./modules/toolbar/better-picker.cjs.js");
const betterToolbar = require("./modules/toolbar/better-toolbar.cjs.js");
const toolbarTip = require("./modules/toolbar/toolbar-tip.cjs.js");
const I18n = require("quill-i18n");
exports.getListValue = index.getListValue;
exports.inputFile = index.inputFile;
exports.namespace = index.namespace;
Object.defineProperty(exports, "generateTableUpShortKeyMenu", {
  enumerable: true,
  get: () => QuillShortcutKey.generateTableUpShortKeyMenu
});
Object.defineProperty(exports, "AttributeMap", {
  enumerable: true,
  get: () => core.AttributeMap
});
Object.defineProperty(exports, "Delta", {
  enumerable: true,
  get: () => core.Delta
});
Object.defineProperty(exports, "Op", {
  enumerable: true,
  get: () => core.Op
});
Object.defineProperty(exports, "OpIterator", {
  enumerable: true,
  get: () => core.OpIterator
});
Object.defineProperty(exports, "Parchment", {
  enumerable: true,
  get: () => core.Parchment
});
Object.defineProperty(exports, "Range", {
  enumerable: true,
  get: () => core.Range
});
exports.default = fluentEditor.default;
exports.SizeStyle = fontSize.SizeStyle;
exports.FontStyle = fontStyle.FontStyle;
exports.LineHeightStyle = lineHeight.LineHeightStyle;
exports.TextIndentStyle = textIndent.TextIndentStyle;
exports.AUDIO_VIDEO_UPLOADER_MIME_TYPES = editor_config.AUDIO_VIDEO_UPLOADER_MIME_TYPES;
exports.BIG_DELTA_LIMIT = editor_config.BIG_DELTA_LIMIT;
exports.CHANGE_LANGUAGE_EVENT = editor_config.CHANGE_LANGUAGE_EVENT;
exports.COMPRESSED_UPLOADER_MIME_TYPES = editor_config.COMPRESSED_UPLOADER_MIME_TYPES;
exports.DEFAULT_TOOLBAR = editor_config.DEFAULT_TOOLBAR;
exports.DOC_UPLOADER_MIME_TYPES = editor_config.DOC_UPLOADER_MIME_TYPES;
exports.FILE_UPLOADER_MIME_TYPES = editor_config.FILE_UPLOADER_MIME_TYPES;
exports.FULL_TOOLBAR = editor_config.FULL_TOOLBAR;
exports.IMAGE_UPLOADER_MIME_TYPES = editor_config.IMAGE_UPLOADER_MIME_TYPES;
exports.OTHER_FILE_UPLOADER_MIME_TYPES = editor_config.OTHER_FILE_UPLOADER_MIME_TYPES;
exports.PPT_UPLOADER_MIME_TYPES = editor_config.PPT_UPLOADER_MIME_TYPES;
exports.XSL_UPLOADER_MIME_TYPES = editor_config.XSL_UPLOADER_MIME_TYPES;
exports.defaultLanguage = editor_config.defaultLanguage;
exports.getEventComposedPath = editor_utils.getEventComposedPath;
exports.hadProtocol = editor_utils.hadProtocol;
exports.hexToRgbA = editor_utils.hexToRgbA;
exports.imageFileToUrl = editor_utils.imageFileToUrl;
exports.imageUrlToFile = editor_utils.imageUrlToFile;
exports.isInside = editor_utils.isInside;
exports.isNullOrUndefined = editor_utils.isNullOrUndefined;
exports.omit = editor_utils.omit;
exports.replaceDeltaImage = editor_utils.replaceDeltaImage;
exports.sanitize = editor_utils.sanitize;
exports.splitWithBreak = editor_utils.splitWithBreak;
exports.EmojiBlot = emoji.EmojiBlot;
exports.SoftBreak = softBreak.SoftBreak;
exports.StrikeBlot = strike.StrikeBlot;
exports.Video = video.Video;
exports.CollaborationModule = module$1.CollaborationModule;
exports.CustomClipboard = customClipboard.CustomClipboard;
exports.Action = action.Action;
exports.CustomResizeAction = customResizeAction.CustomResizeAction;
exports.DeleteAction = deleteAction.DeleteAction;
exports.ALIGN_ATTR = imageToolbarButtons.ALIGN_ATTR;
exports.ImageToolbarButtons = imageToolbarButtons.ImageToolbarButtons;
exports.alignmentHandler = imageToolbarButtons.alignmentHandler;
exports.setAlignStyle = imageToolbarButtons.setAlignStyle;
exports.ImageToolbar = toolbar.ImageToolbar;
exports.ImageToolbarAction = toolbarAction.ImageToolbarAction;
exports.BlotFormatter = blotFormatter.BlotFormatter;
exports.CustomImage = image.CustomImage;
exports.ImagePreviewModal = previewModal.ImagePreviewModal;
exports.getImagePreviewModal = previewModal.getImagePreviewModal;
exports.BlotSpec = blotSpec.BlotSpec;
exports.CustomImageSpec = customImageSpec.CustomImageSpec;
exports.ImageSpec = imageSpec.ImageSpec;
exports.FileUploader = customUploader.FileUploader;
exports.DividerBlot = divider.DividerBlot;
exports.EmojiModule = emoji$1.EmojiModule;
exports.File = file.File;
exports.FileBar = fileBar.FileBar;
exports.FileModule = fileModule.FileModule;
exports.FlowChartModule = index$1.FlowChartModule;
exports.LinkBlot = link.LinkBlot;
exports.LinkTooltip = tooltip.LinkTooltip;
exports.MathliveBlot = formats.MathliveBlot;
exports.MathliveModule = module$2.MathliveModule;
exports.MathliveTooltip = tooltip$1.MathliveTooltip;
exports.Mention = mention.Mention;
exports.MentionLink = mentionLink.MentionLink;
exports.MindMapModule = index$2.MindMapModule;
exports.ShortCutKey = index$3.ShortCutKey;
exports.shortKey = index$3.shortKey;
exports.generateTableUp = index$4.generateTableUp;
exports.ColorPicker = betterPicker.ColorPicker;
exports.Picker = betterPicker.Picker;
exports.BetterToolbar = betterToolbar.BetterToolbar;
exports.generateToolbarTip = toolbarTip.generateToolbarTip;
Object.keys(QuillShortcutKey).forEach((k) => {
  if (k !== "default" && !Object.prototype.hasOwnProperty.call(exports, k)) Object.defineProperty(exports, k, {
    enumerable: true,
    get: () => QuillShortcutKey[k]
  });
});
Object.keys(I18n).forEach((k) => {
  if (k !== "default" && !Object.prototype.hasOwnProperty.call(exports, k)) Object.defineProperty(exports, k, {
    enumerable: true,
    get: () => I18n[k]
  });
});
